code/modules/paperwork/photocopier.dm

```diff
@@ -97,35 +117,73 @@ GLOBAL_LIST_INIT(paper_blanks, init_paper_blanks())
 	toner_cartridge = new(src)
 	return ..()
 
+/obj/machinery/photocopier/get_save_vars()
+	. = ..()
+	. += NAMEOF(src, paper_stack)
+	return .
+
 /obj/machinery/photocopier/Initialize(mapload)
 	. = ..()
-	// Creates the paper and inserts it so we can drop it when we get destroyed/deconstructed
-	for(var/i in 1 to starting_paper)
-		paper_stack += new created_paper(src)
+	RefreshParts()
 	setup_components()
 	AddElement(/datum/element/elevation, pixel_shift = 8) //enough to look like your bums are on the machine.
+	if(starting_paper)
+		paper_stack[created_paper] = starting_paper
 
 /// Simply adds the necessary components for this to function.
 /obj/machinery/photocopier/proc/setup_components()
 	AddComponent(/datum/component/payment, PHOTOCOPIER_FEE, SSeconomy.get_dep_account(ACCOUNT_CIV), PAYMENT_CLINICAL)
 
+/obj/machinery/photocopier/RefreshParts()
+	. = ..()
+	max_paper_capacity = 0
+	for(var/datum/stock_part/matter_bin/matter_bin in component_parts)
+		max_paper_capacity += MATTER_BIN_PAPER_CAPACITY * matter_bin.tier
+
+	toner_efficiency = 1
+	for(var/datum/stock_part/micro_laser/micro_laser in component_parts)
+		toner_efficiency += micro_laser.tier
+
+	time_to_print = 5 SECONDS
+	for(var/datum/stock_part/scanning_module/scanning_module in component_parts)
+		time_to_print -= (scanning_module.tier SECONDS)
```

Comment:
you iterate same list 3 times, instead of doing typechecks during 1 iteration like this:
```
max_paper_capacity = 0
toner_efficiency = 1
time_to_print = 5 SECONDS
for(var/datum/stock_part/part as anything in component_parts)
	if(istype(part, /datum/stock_part/matter_bin)
		max_paper_capacity += MATTER_BIN_PAPER_CAPACITY * matter_bin.tier
	else if(istype(part, /datum/stock_part/micro_laser)
		toner_efficiency += micro_laser.tier
	else if(istype(part, /datum/stock_part/scanning_module)
		time_to_print -= (scanning_module.tier SECONDS)
```

Reply:
Good catch. I actually reused this code from other machines RefreshParts() proc so it can probably be applied in many places.

Reply:
> you iterate same list 3 times, instead of doing typechecks during 1 iteration like this:

![chrome_Ef0WEGFPmI](https://github.com/user-attachments/assets/632289df-dc52-4ecf-bcc9-8f8164a6132b)

Just a heads up, you can use the "Add a suggestion" button to suggest code edits that will get commited to the PR and attribute you as a co-author. Your code in the current form is broken though because you need to do `part.tier` instead of `matter_bin.tier`, `micro_laser.tier`, `scanning_module.tier`

Reply:
Looks like this is a bad idea:
- https://github.com/tgstation/tgstation/pull/91301